byoc/trickle.go

```diff
@@ -0,0 +1,952 @@
+					return
+				}
+				continue
+			}
+
+			retries = 0
+
+			body, err := io.ReadAll(segment.Body)
+			segment.Body.Close()
+			if err != nil {
+				clog.Infof(ctx, "Error reading events subscription body: %s", err)
+				continue
+			}
+
+			var eventWrapper struct {
+				QueueEventType string                 `json:"queue_event_type"`
+				Event          map[string]interface{} `json:"event"`
+			}
+			if err := json.Unmarshal(body, &eventWrapper); err != nil {
+				clog.Infof(ctx, "Failed to parse JSON from events subscription: %s", err)
+				continue
+			}
+
+			event := eventWrapper.Event
+			queueEventType := eventWrapper.QueueEventType
+			if event == nil {
+				event = make(map[string]interface{})
+				if err := json.Unmarshal(body, &event); err != nil {
+					clog.Infof(ctx, "Failed to parse JSON as direct event: %s", err)
+					continue
+				}
+				queueEventType = "ai_stream_events"
+			}
+
+			event["stream_id"] = streamId
+			event["request_id"] = params.liveParams.requestID
+			event["pipeline_id"] = params.liveParams.pipelineID
+			event["orchestrator_info"] = map[string]interface{}{
+				"address": orchAddr,
+				"url":     orchUrl,
+			}
+
+			clog.V(8).Infof(ctx, "Received event for seq=%d event=%+v",
+				trickle.GetSeq(segment), event,
+			)
+
+			lastEventMu.Lock()
+			lastEvent = time.Now()
+			lastEventMu.Unlock()
+
+			eventType, ok := event["type"].(string)
+			if !ok {
+				eventType = "unknown"
+				clog.Warningf(ctx, "Received event without a type stream=%s event=%+v",
+					streamId, event,
+				)
+			}
+
+			if eventType == "status" {
+				queueEventType = "ai_stream_status"
+
+				lastStreamStatus, _ := bsg.statusStore.Get(streamId)
+				inferenceStatus, hasInference := event["inference_status"].(map[string]interface{})
+				lastInferenceStatus, hasLastInference := lastStreamStatus["inference_status"].(map[string]interface{})
+
+				if hasInference {
+					if logs, ok := inferenceStatus["last_restart_logs"]; !ok || logs == nil {
+						if hasLastInference {
+							inferenceStatus["last_restart_logs"] =
+								lastInferenceStatus["last_restart_logs"]
+						}
+					}
+					if params, ok := inferenceStatus["last_params"]; !ok || params == nil {
+						if hasLastInference {
+							inferenceStatus["last_params"] =
+								lastInferenceStatus["last_params"]
+						}
+					}
+				}
+
+				bsg.statusStore.Store(streamId, event)
+			}
+
+			monitor.SendQueueEventAsync(queueEventType, event)
+		}
+	}()
+
+	// Heartbeat watchdog
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-eventsDone:
+				return
+			case <-eventTicker.C:
+				lastEventMu.Lock()
+				eventTime := lastEvent
+				lastEventMu.Unlock()
+
+				if time.Since(eventTime) > maxEventGap {
+					stopProcessing(ctx, params, fmt.Errorf("timeout waiting for events"))
+					return
+				}
+			}
+		}
+	}()
+}
+
+func (bsg *BYOCGatewayServer) getOutWriter(streamId string) (*media.RingBuffer, string) {
+	bsg.mu.Lock()
```

Comment:
Looks like getOutWriter grabs bsg.mu and then calls streamPipeline, which also locks bsg.mu, so we deadlock before waiting on OutCond. Since streamPipeline already locks, could we drop the outer lock here and keep a single critical section for the cond wait?
```suggestion
func (bsg *BYOCGatewayServer) getOutWriter(streamId string) (*media.RingBuffer, string) {
	stream, err := bsg.streamPipeline(streamId) // streamPipeline handles locking
	if err != nil || stream.Closed {
		return nil, ""
	}
	// hold the cond lock only while waiting
	stream.OutCond.L.Lock()
	defer stream.OutCond.L.Unlock()

	for stream.OutWriter == nil {
		stream.OutCond.Wait()
		if stream.Closed {
			return nil, ""
		}
	}
	return stream.OutWriter, stream.RequestID
}
```